internal/controller/gatewayclass.go

```diff
@@ -0,0 +1,127 @@
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+	gatewayapi_v1alpha1 "sigs.k8s.io/gateway-api/apis/v1alpha1"
+)
+
+type gatewayClassReconciler struct {
+	client       client.Client
+	eventHandler cache.ResourceEventHandler
+	log          logrus.FieldLogger
+	className    string
+}
+
+// NewGatewayClassController creates the gatewayclass controller. The controller
+// will be pre-configured to watch for cluster-scoped GatewayClass objects with
+// a controller field that matches name.
+func NewGatewayClassController(mgr manager.Manager, eventHandler cache.ResourceEventHandler, log logrus.FieldLogger, name string) (controller.Controller, error) {
+	r := &gatewayClassReconciler{
+		client:       mgr.GetClient(),
+		eventHandler: eventHandler,
+		log:          log,
+		className:    name,
+	}
+
+	c, err := controller.New("gatewayclass-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return nil, err
+	}
+
+	// Only enqueue GatewayClass objects that match name.
+	if err := c.Watch(&source.Kind{Type: &gatewayapi_v1alpha1.GatewayClass{}}, r.enqueueRequestForGatewayClass(name)); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// enqueueRequestForGatewayClass returns an event handler that maps events to
+// GatewayClass objects with a controller field that matches name.
+func (r *gatewayClassReconciler) enqueueRequestForGatewayClass(name string) handler.EventHandler {
+	return handler.EnqueueRequestsFromMapFunc(func(a client.Object) []reconcile.Request {
+		gc, ok := a.(*gatewayapi_v1alpha1.GatewayClass)
+		if !ok {
+			r.log.WithField("name", a.GetName()).Info("invalid object, bypassing reconciliation.")
+			return []reconcile.Request{}
+		}
+		if gc.Spec.Controller == name {
+			r.log.WithField("name", gc.Name).Info("queueing gatewayclass")
+			return []reconcile.Request{
+				{
+					NamespacedName: types.NamespacedName{
+						Name: gc.Name,
+					},
+				},
+			}
+		}
+		r.log.WithField("name", gc.Name).Info("controller is not ", name, "; bypassing reconciliation")
+		return []reconcile.Request{}
+	})
+}
+
+func (r *gatewayClassReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	r.log.WithField("name", request.Name).Info("reconciling gatewayclass")
+
+	// Fetch the Gateway from the cache.
+	gc := &gatewayapi_v1alpha1.GatewayClass{}
+	if err := r.client.Get(ctx, request.NamespacedName, gc); err != nil {
+		if api_errors.IsNotFound(err) {
+			r.log.WithField("name", request.Name).Info("failed to find gatewayclass")
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object, so requeue the request.
+		return reconcile.Result{}, fmt.Errorf("failed to get gatewayclass %q: %w", request.Name, err)
+	}
+
+	// Check if object is marked for deletion.
+	if !gc.ObjectMeta.DeletionTimestamp.IsZero() {
+		r.eventHandler.OnDelete(gc)
```

Comment:
A future issue we should discuss what to do when the class is deleted. Should we drop all configurations from Envoy? Throw an error? This would be super descructive. =)

Reply:
xref: https://github.com/projectcontour/contour/issues/3700